feat: Region pinning support

typester opened the pull request.

bcherry commented on livekit/src/rtc_engine/mod.rs:
```diff
@@ -337,13 +337,15 @@ impl EngineInner {
         let lk_runtime = LkRuntime::instance();
         let max_retries = options.join_retries;
 
+        let urls = RegionUrlProvider::fetch_region_urls(url, token).await?;
```
what is the precise behavior if region pinning is turned on for the project but you just connect to the default URL? does it route correctly under the hood anyways, or does it route to nearest but fail if that's a disallowed region?

typester replied:
Ok, I will update the code.

theomonnom commented on livekit/src/rtc_engine/mod.rs:
```diff
@@ -337,13 +337,15 @@ impl EngineInner {
         let lk_runtime = LkRuntime::instance();
         let max_retries = options.join_retries;
 
+        let urls = RegionUrlProvider::fetch_region_urls(url, token).await?;
```
Should we use the RegionUrlProvider directly inside the SignalClient instead? in the livekit-api crate

theomonnom replied:
Some people use the SignalClient but doesn't use our `livekit` crate

typester replied:
That makes sense. Will do it

typester commented on livekit/src/rtc_engine/mod.rs:
```diff
@@ -379,20 +381,22 @@ impl EngineInner {
+                    Err(e) => {
+                        let attempt_i = i + 1;
+                        if i < max_retries {
+                            log::warn!(
```
Oh yeah, I thought the same thing but I forgot to update the code somehow :sweat_smile: 
Will do

typester commented:
I made the following changes:

- Moved the connection fallback to regional URLs functionality to the SignalClient.
- Removed the initial fetching of regional URLs. If the default connection is refused, try connecting to the regional URLs.
- Log an error if the default URL returns an error other than 403.

davidzhao commented on livekit-api/src/signal_client/region.rs:
```diff
@@ -0,0 +1,70 @@
+                .map_err(|e| SignalError::RegionError(e.to_string()))?;
+            Ok(res.regions.into_iter().map(|i| i.url).collect())
+        } else {
+            Ok(vec![url.into()])
```
IMO this should be returning empty.. no regional urls are fetched

typester replied:
Oh, yes. Will fix

lukasIO reviewed (approved):
lgtm, the only nit I have is that now for region pinned projects users that don't connect correctly via geo always get the penalty of a failed connection attempt first. 

Ideally we'd offer a `prepare_connection` method on top that sets the region url for the next connect attempt. see https://github.com/livekit/client-sdk-js/blob/main/src/room/Room.ts#L656-L665

typester commented:
> lgtm, the only nit I have is that now for region pinned projects users that don't connect correctly via geo always get the penalty of a failed connection attempt first.
> 
> Ideally we'd offer a `prepare_connection` method on top that sets the region url for the next connect attempt. see https://github.com/livekit/client-sdk-js/blob/main/src/room/Room.ts#L656-L665

Yeah, we'll need this at some point in the future, but I think it's fine for this PR in terms of minimal implementation.
I'll keep it in mind. Thank you!
